spark/src/main/java/org/apache/iceberg/spark/source/Reader.java

```diff
@@ -396,48 +397,64 @@ public void close() throws IOException {
       // schema or rows returned by readers
       Schema finalSchema = expectedSchema;
       PartitionSpec spec = task.spec();
-      Set<Integer> idColumns = spec.identitySourceIds();
+      Set<Integer> idColumns = Sets.newHashSet();
+      for (Integer i : spec.identitySourceIds()) {
+        if (spec.schema().columns().stream()
+            .noneMatch(j -> j.type().isStructType() && j.type().asStructType().field(i) != null)
+        ) {
+          idColumns.add(i);
+        }
+      }
```

Comment:
Why is this still running a check on identity partitions? The partition value map should include all identity partition values and they will be added in the right place. Fields that aren't projected will be ignored.